src/store_client.cc

```diff
@@ -44,6 +45,41 @@ static bool CheckQuickAbortIsReasonable(StoreEntry * entry);
 
 CBDATA_CLASS_INIT(store_client);
 
+
+/* StoreClient */
+
+bool
+StoreClient::mayInitiateCollapsing() const
+{
+    if (!Config.onoff.collapsed_forwarding)
+        return false;
+
+    if (!Config.accessList.collapsedForwardingAccess)
+        return true;
+
+    ACLFilledChecklist checklist(Config.accessList.collapsedForwardingAccess, nullptr, nullptr);
+    fillChecklist(checklist);
+    return checklist.fastCheck().allowed();
+}
+
+bool
+StoreClient::mayCollapseOn(const StoreEntry&) const
+{
+    // XXX: Implement. Very similar to mayInitiateCollapsing() but also checks
+    // whether we are about to become a collapsed slave (and not just a regular
+    // hit): For regular hits, returns true. For others, checks the new ACL.
```

Comment:
This looks a little strange because we are not going to collapse for regular hits but still return true.

Reply:
I agree that there is a problem with the current mayCollapseOn() name. As you noticed, the method has to return true for regular hits but the method name does not reflect that. This is one of the polishing things I ran out of time to do last night.

There are two ways to fix that: find a better name/definition (see below) or split the method (your suggestion). I am not sure we should duplicate `e.cfInitiator() && !mayCollapse()` expressions/logic in ten places, but finding a good name that reflects the "and" condition may also be impossible. There is no a clear winner in this design choice IMO.

For the naming option: We may want to reverse the method polarity to find that better name. The method essence is not about whether collapsing on the entry is allowed; it is rather whether collapsing on the entry is possible but prohibited. We could use something like cfEnslavementProhibited(e) instead. It will (almost naturally) return false for regular hits because regular hits do not prohibit CF enslavement. I said "almost" because regular hits do not allow CF enslavement either.

On duplicating `e.cfInitiator() && !mayCollapse()` logic: The biggest problem here is that somebody can accidentally call mayCollapse() without the cfInitiator() guard. It would be difficult to spot such bugs visually. If you decide to use this option, please keep two methods and assert cfInitiator() in mayCollapseOn(e). Testing cfInitiator() twice is a bit wasteful, but I think it is not prohibitively expensive:

```
bool
StoreClient::mayInitiateCollapsing() const
{
    return onCollapsingPath();
}

bool
StoreClient::mayCollapseOn(const StoreEntry &e) const
{
    assert(e.cfInitator()); // our result is not meaningful for regular hits
    return onCollapsingPath();
}
```

Pick whatever option (renaming or splitting) you think is the best.

Adding something like StoreEntry::collapsingInitiator() and StoreClient::onCollapsingPath() is probably a good idea in either case. I was thinking about doing that but ran out of time.

Reply:
> e.cfInitiator() && !mayCollapse() logic

I used this approach, it looks more clear even than using `cfEnslavementProhibited(e)`: this method it should act differently for two (similar otherwise) 'prohibit' and 'do not allow' notions, returning 'true' in the first case and 'false' in the second.